lib/inputs/dovetail-frequency.js

```diff
@@ -0,0 +1,198 @@
+      logger.warn(msg, { ddb: 'retrying', count: failureCount });
+      const err = new Error(msg);
+      err.skipLogging = true;
+      throw err;
+    }
+
+    return output;
+  }
+
+  // spin up N workers, and process the queue of updates
+  async updateAll(updates, concurrency = 25) {
+    let successCount = 0;
+    let deleteCount = 0;
+    let failureCount = 0;
+
+    const client = await dynamo.client();
+    const worker = async () => {
+      let args;
+      while ((args = updates.shift())) {
+        try {
+          const params = this.updateParams(args);
+          const result = await dynamo.updateItemPromise(params, client);
+
+          const toDelete = this.evictImpressions(args, result);
+          if (toDelete.length > 0) {
+            const deleteParams = this.deleteParams(args, toDelete);
+            const deleteResult = await dynamo.updateItemPromise(deleteParams, client);
+            deleteCount++;
+          }
+
+          successCount++;
+        } catch (err) {
+          if (err.name === 'ProvisionedThroughputExceededException') {
+            logger.warn(`DDB throughput exceeded [${process.env.DDB_FREQUENCY_TABLE}]: ${err}`, { args });
+          } else {
+            logger.error(`DDB Error [${process.env.DDB_FREQUENCY_TABLE}]: ${err}`, { args });
+          }
+          failureCount++;
+        }
+      }
+    };
+
+    const threads = Array(concurrency).fill(true);
+    await Promise.all(threads.map(() => worker()));
+    return [successCount, deleteCount, failureCount];
+  }
+
+  minImpressionTime(data) {
+    return Math.floor((new Date().getTime() / 1000) - data.maxSeconds);
+  }
+
+  evictImpressions(data, result) {
+    let toEvict = [];
+    if (result.Attributes && result.Attributes.impressions) {
+      const impressions = result.Attributes.impressions.NS || [];
+      const evictTime = this.minImpressionTime(data);
+      const evictList = impressions.filter((imp) => imp <= evictTime);
+
+      // Only evict if the list is half or more of the total list, and greater than 10
+      if (evictList.length > 10 && evictList.length * 2 >= impressions.length) {
+        toEvict = evictList;
+      }
+    }
+    return toEvict;
+  }
+
+  deleteParams(data, deleteArray) {
+    return {
+      TableName: this.tableName(),
+      Key: {
+        listener: { S: `${data.listener}` },
+        campaign: { S: `${data.campaign}` },
+      },
+      ReturnValues: 'ALL_NEW',
+      UpdateExpression: "DELETE impressions :td",
+      ExpressionAttributeValues: {
+        ":td": { "NS": deleteArray.map((td) => `${td}`) },
+      },
+    };
+  }
+
+  updateParams(data) {
+    return {
+      TableName: this.tableName(),
+      Key: {
+        listener: { S: `${data.listener}` },
+        campaign: { S: `${data.campaign}` },
+      },
+      ReturnValues: 'ALL_OLD',
+      UpdateExpression: "SET expiration = :ttl ADD impressions :ts",
```

Comment:
So you say that retrying this batch of kinesis records would be idempotent...

But it seems like we'd end up `ADD`-ing the same timestamp twice, right?  Does DTR filter those duplicate timestamps out then?

(I should probably just go look...)

Reply:
it is a set in dynamodb, if you add the timestamp again, DDB doesn't add it again, unique timestamps only.

Reply:
Here's my output from doing this against staging - I keep calling ADD with the same data, but the set doesn't update or error:
https://gist.github.com/kookster/25628f1a49c2ee82672c8bbfe7ea4c91

I also described more about this in... the RFC!
https://github.com/PRX/internal/blob/main/rfcs/0028_frequency_capping.md#dynamo-implementation

Reply:
Basically, I picked this particular way to use DDB as my research indicated this was the best way to get an idempotent list of values, as I didn't want to have to manage deduping in our code.

Reply:
Ah, gotcha.  Didn't realize that field was a Set (not an Array).  +1